feat(isthmus): define `OpExecutionData`

emhane opened the pull request:
Closes https://github.com/alloy-rs/op-alloy/issues/417

- Defines `OpExecutionData` which wraps `OpExecutionPayload` and `OpExecutionPayloadSidecar`
- Debugs `OpExecutionPayload`
- Removes `OpExecutionPayloadV4`, since the l2 withdrawals root goes in sidecar, it doesn't add an extra field to the header

mattsse commented on crates/rpc-types-engine/src/payload/mod.rs:
```diff
@@ -3,19 +3,201 @@
     V2(ExecutionPayloadV2),
     /// V3 payload
     V3(ExecutionPayloadV3),
-    /// V4 payload
-    V4(OpExecutionPayloadV4),
```
why is this removed?

emhane replied:
ye

emhane replied:
see https://specs.optimism.io/protocol/isthmus/exec-engine.html#update-to-executabledata

mattsse replied:
still unclear what the parameters are, from reading the spec I'd assume that the withdrawals root is an additional attribute of the payload param and not a standalone parameter?

emhane replied:
yes, we know, we are working on it

mattsse reviewed (commented):
> Removes OpExecutionPayloadV4, since the l2 withdrawals root goes in sidecar, it doesn't add an extra field to the header

I'm very confused by this now, 
so the newPayloadV4 endpoint gets an additional argument and the payload object remains unchanged?

mattsse commented on crates/rpc-types-engine/src/payload/v4.rs:
```diff
@@ -4,21 +4,6 @@ use alloc::vec::Vec;
-    /// L1 execution payload
-    #[cfg_attr(feature = "serde", serde(flatten))]
-    pub payload_inner: ExecutionPayloadV3,
-    /// OP-Stack Isthmus specific field:
-    /// instead of computing the root from a withdrawals list, set it directly.
-    /// The "withdrawals" list attribute must be non-nil but empty.
-    pub withdrawals_root: B256,
```
how are we getting this field over rpc now?

emhane replied:
always as envelope afaik

mattsse replied:
the envelope is what `getPayload` returns, this type is the raw payload

I know this is confusing

mattsse reviewed (commented):
this is a bit confusing, how does the intended flow look like from rpc pov?

the isthmuspayloadfields(withdrawalsroot) actually doesn't make much sense because we need to capture this as part of the payload over rpc.

so are we doing
parse OpPayloadV4 on newPayload, split then split that into V3 and Isthmus payloadfields?

refcell commented on crates/rpc-types-engine/src/envelope.rs:
```diff
@@ -24,32 +74,7 @@ pub struct OpNetworkPayloadEnvelope {
 
 impl OpNetworkPayloadEnvelope {
     /// Decode a payload envelope from a snappy-compressed byte array.
-    /// The payload version decoded is `ExecutionPayloadV1` from SSZ bytes.
-    #[cfg(feature = "std")]
-    pub fn decode_v1(data: &[u8]) -> Result<Self, PayloadEnvelopeError> {
```
Dumb Q - but why was this removed?

emhane replied:
because op doesn't support v1 https://specs.optimism.io/protocol/exec-engine.html can be out of scope to fix this semantic in this pr

refcell reviewed (commented):
Would it make sense to split the changes requested in #417 into it's own PR as [the upstream alloy pr did](https://github.com/alloy-rs/alloy/pull/2003) for ethereum types?

This way we can split out the endpoint changes which is causing confusion.

mattsse commented:
> type into it's own PR without modifying other code

makes sense,

I'll summarize the core issue with this PR:

* with isthmus opstack breaks L1 compatibility with the payload parameter of the `newPayloadV4([payload, (sidecar)])` endpoint.
* because we need to parse `payload` as an object, we need a dedicated opstack compatible ExecutionPayload type (this type is currently named `OpExecutionPayloadV4`)
* We currently have `IsthumusPayloadFields` for the withdrawals_root but this is already part of the payload object, so this type seems redundant

The problem from reth POV
* because newPayload does not accepts multiple param objects, not a single object we need to add the helper type `ExecutionData` so that we can easily handle arbitrary execution requests, this is effectively extra work necessary due to engine API design

emhane commented:
> Would it make sense to split the changes requested in #417 into it's own PR as [the upstream alloy pr did](https://github.com/alloy-rs/alloy/pull/2003) for ethereum types? That is, specifically split the `OpExecutionData` type into it's own PR without modifying other code.
> 
> This way we can split out the endpoint changes which is causing confusion.

done, removed the impl body of the new type that i had implemented to match the alloy type `ExecutionData` impl, since this was what required me to change parts of existing code

emhane commented:
> > type into it's own PR without modifying other code
> 
> makes sense,
> 
> I'll summarize the core issue with this PR:
> 
> * with isthmus opstack breaks L1 compatibility with the payload parameter of the `newPayloadV4([payload, (sidecar)])` endpoint.
> * because we need to parse `payload` as an object, we need a dedicated opstack compatible ExecutionPayload type (this type is currently named `OpExecutionPayloadV4`)
> * We currently have `IsthumusPayloadFields` for the withdrawals_root but this is already part of the payload object, so this type seems redundant
> 
> The problem from reth POV
> 
> * because newPayload does not accepts multiple param objects, not a single object we need to add the helper type `ExecutionData` so that we can easily handle arbitrary execution requests, this is effectively extra work necessary due to engine API design

all great points that I'm glad surface from this pr. makes sense to open a separate issues.
